### PR TITLE
Sparse Mixture-of-Experts FFN layers in last Transolver blocks

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replacing the dense FFN in the last 1-2 Transolver blocks with a Sparse Mixture-of-Experts (MoE) FFN will increase model capacity without proportionally increasing compute at inference. Different experts can specialize in different flow regimes (attached flow vs. separated, near-wall vs. far-field), improving generalization across geometries. With n_experts=8 and top-2 routing, the effective capacity roughly 4x's the FFN without changing the per-token FLOP count significantly.

**Cross-dataset coverage:** TandemFoil (TF), TandemFoil Paper (TFP), AirfRANS (AF), DrivAerML (DM).

## Current Baselines

| Dataset | Metric | Value | PR |
|---------|--------|-------|----|
| TF | `val_primary/surface_pressure_mae` | 22.537 | #2924 |
| TFP | `val_primary/field_mse` | 0.00434 | #2979 |
| AF | `val_primary/surface_mse` | 0.000340 | #2898 |
| DM | `val_primary/surface_rel_l2_pct` | 3.997% | #2898 |

**External DrivAerML target:** <3.71% (AB-UPT benchmark)

## Implementation

Add `--moe-layers`, `--moe-n-experts`, and `--moe-load-balance-weight` flags to `train.py`. Modify the Transolver architecture to replace the FFN in the last `--moe-layers` blocks with a `SparseMoEFFN`.

**SparseMoEFFN architecture:**
```python
class SparseMoEFFN(nn.Module):
    def __init__(self, d_model, d_ff, n_experts=8, top_k=2):
        super().__init__()
        self.n_experts = n_experts
        self.top_k = top_k
        self.router = nn.Linear(d_model, n_experts, bias=False)
        self.experts = nn.ModuleList([
            nn.Sequential(
                nn.Linear(d_model, d_ff),
                nn.GELU(),
                nn.Linear(d_ff, d_model)
            ) for _ in range(n_experts)
        ])
    
    def forward(self, x):
        # x: [B, N, D]
        router_logits = self.router(x)  # [B, N, E]
        routing_weights, selected_experts = torch.topk(
            torch.softmax(router_logits, dim=-1), self.top_k, dim=-1
        )  # [B, N, top_k]
        routing_weights = routing_weights / routing_weights.sum(dim=-1, keepdim=True)
        
        output = torch.zeros_like(x)
        for k in range(self.top_k):
            expert_idx = selected_experts[..., k]  # [B, N]
            weight = routing_weights[..., k:k+1]   # [B, N, 1]
            for e in range(self.n_experts):
                mask = (expert_idx == e)
                if mask.any():
                    output[mask] += weight[mask] * self.experts[e](x[mask])
        return output
    
    def load_balance_loss(self, x):
        router_logits = self.router(x)
        probs = torch.softmax(router_logits, dim=-1)  # [B, N, E]
        # Encourage uniform expert utilization
        mean_prob = probs.mean(dim=(0, 1))  # [E]
        return self.n_experts * (mean_prob * torch.log(mean_prob + 1e-8)).sum()
```

**In the training loop, add auxiliary loss:**
```python
moe_lb_loss = sum(layer.ffn.load_balance_loss(x) for layer in moe_layers)
total_loss = data_loss + args.moe_load_balance_weight * moe_lb_loss
```

**Apply MoE to the last `--moe-layers` Transolver blocks only.** Start with 1 (just the final block) to minimize risk.

## Experiment Instructions

Run the following experiments (GPU allocation: DM=4 GPUs, TF=2 GPUs, AF=1 GPU, TFP=1 GPU):

**TandemFoil (2 GPUs):**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --moe-layers 1 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --wandb-group moe-ffn-layers
```

**TandemFoil Paper (1 GPU):**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 150 --no-use-ema \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --moe-layers 1 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --wandb-group moe-ffn-layers
```

**AirfRANS (1 GPU):**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --moe-layers 1 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --wandb-group moe-ffn-layers
```

**DrivAerML (4 GPUs):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --moe-layers 1 --moe-n-experts 8 --moe-load-balance-weight 0.01 \
  --wandb-group moe-ffn-layers
```

## Expected Outcome

- DM: improvement from 3.997% toward <3.71% (specialized experts for different aerodynamic regimes: attached flow, separated flow, wake)
- TF: improvement from 22.537 (expert specialization for upstream vs downstream foil near-wake regions)
- AF: improvement from 0.000340 (boundary layer vs freestream specialization)

## Suggested Follow-ups

- If `--moe-layers 1` helps, try `--moe-layers 2` for the last two blocks
- Ablate number of experts: n_experts=4 vs 8 vs 16
- If load balance loss is too high (experts collapse to 1-2), increase `moe_load_balance_weight` to 0.1
- Monitor expert utilization via router entropy — log `mean_prob.entropy()` as a W&B metric

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
